ddl/ddl_tiflash_api.go

```diff
@@ -0,0 +1,504 @@
+	}
+}
+
+// UpdateTiFlashHTTPAddress report TiFlash's StatusAddress's port to Pd's etcd.
+func (d *ddl) UpdateTiFlashHTTPAddress(store *helper.StoreStat) error {
+	addrAndPort := strings.Split(store.Store.StatusAddress, ":")
+	if len(addrAndPort) < 2 {
+		return errors.New("Can't get TiFlash Address from PD")
+	}
+	httpAddr, err := getTiflashHTTPAddr(addrAndPort[0], store.Store.StatusAddress)
+	if err != nil {
+		return errors.Trace(err)
+	}
+	// Report to pd
+	key := fmt.Sprintf("/tiflash/cluster/http_port/%v", store.Store.Address)
+	if d.etcdCli == nil {
+		return errors.New("no etcdCli in ddl")
+	}
+	origin := ""
+	resp, err := d.etcdCli.Get(d.ctx, key)
+	if err != nil {
+		// Try to update.
+		for _, kv := range resp.Kvs {
+			if string(kv.Key) == key {
+				origin = string(kv.Value)
+				break
+			}
+		}
+	}
+	if origin != httpAddr {
+		logutil.BgLogger().Warn(fmt.Sprintf("Update status addr to %v", httpAddr))
+		err := ddlutil.PutKVToEtcd(d.ctx, d.etcdCli, 1, key, httpAddr)
+		if err != nil {
+			return errors.Trace(err)
+		}
+	}
+
+	return nil
+}
+
+func updateTiFlashStores(tikvHelper *helper.Helper, pollTiFlashContext *TiFlashManagementContext) error {
+	// We need the up-to-date information about TiFlash stores.
+	// Since TiFlash Replica synchronize may happen immediately after new TiFlash stores are added.
+	tikvStats, err := tikvHelper.GetStoresStat()
+	if err != nil {
+		return errors.Trace(err)
+	}
+	for _, store := range tikvStats.Stores {
+		for _, l := range store.Store.Labels {
+			if l.Key == "engine" && l.Value == "tiflash" {
+				pollTiFlashContext.TiFlashStores[store.Store.ID] = store
+				logutil.BgLogger().Debug("Find tiflash store", zap.Int64("id", store.Store.ID), zap.String("Address", store.Store.Address), zap.String("StatusAddress", store.Store.StatusAddress))
+			}
+		}
+	}
+	return nil
+}
+
+func (d *ddl) PollTiFlashReplicaStatus(ctx sessionctx.Context, pollTiFlashContext *TiFlashManagementContext) (bool, error) {
+	allReplicaReady := true
+	defer func() {
+		pollTiFlashContext.HandlePdCounter += 1
+		pollTiFlashContext.HandlePdCounter %= PullTiFlashPdTick
+	}()
+
+	tikvStore, ok := ctx.GetStore().(helper.Storage)
+	if !ok {
+		return false, errors.New("Can not get Helper")
+	}
+	tikvHelper := &helper.Helper{
+		Store:       tikvStore,
+		RegionCache: tikvStore.GetRegionCache(),
+	}
+
+	updateTiFlash := pollTiFlashContext.UpdateTiFlashStoreCounter%UpdateTiFlashStoreTick == 0
+	if updateTiFlash {
+		if err := updateTiFlashStores(tikvHelper, pollTiFlashContext); err != nil {
+			// If we failed to get from pd, retry everytime.
+			pollTiFlashContext.UpdateTiFlashStoreCounter = 0
+			return false, errors.Trace(err)
+		}
+		pollTiFlashContext.UpdateTiFlashStoreCounter += 1
+		pollTiFlashContext.UpdateTiFlashStoreCounter %= UpdateTiFlashStoreTick
+	}
+
+	// The following loop updates TiFlash store's status address.
+	for _, store := range pollTiFlashContext.TiFlashStores {
```

Comment:
UpdateTiFlashHTTPAddress is to update TiFlash's status port to PD.
This can be removed currently, since it is a part of old Cluster Manager.
However, TiFlash is currently refactor and will not rely on this later, so it can be removed if it is longer relied by other components, after TiFlash's refactor.